clean up duplicate

ngqinzhe opened the pull request:
My intention was to clean up the duplicate code, however, previous pr was very careless. 

Addressed the issues and heeded your advice, builds successfully as well this time round.

MatchResultUpdate should replace the duplicate code and return the updated matchResult. 

Let me know if there can be further improvements

Tzadiko commented on src/OrderbookCS/MatchingAlgorithm/LifoMatchingAlgorithm.cs:
```diff
@@ -62,11 +62,7 @@ public MatchResult Match(IEnumerable<OrderbookEntry> bids, IEnumerable<Orderbook
                 // Think of refactoring this by including it elsewhere
                 var tradeResult = TradeUtilities.CreateTradeAndFills(orderToMatchBid.Current, orderToMatchAsk.Current,
```
Better, but pass in `FillQuantity` and `AllocationAlgorithm` to `Update`. Create `tradeResult` in `MathResultUpdate`'s `Update`

ngqinzhe replied:
If `tradeResult` is created in `MatchResultCreator` then the conditional checks for `tradeResult.BuyFill.IsCompleteFill` and `tradeResult.SellFill.IsCompleteFill` will have to be placed inside `MatchResultCreator`. Is that the approach you are going for?

Tzadiko replied:
You'll probably need to rethink this. The problems stems from `matchResult` being created at the top of the function IMO. I don't like taking in a `matchResult` and returning a `matchResult` in `CreateMatchResult`, even though that's what I said you should do. The method is called `CreateMatchResult` so it should take in things needed to create a `MatchResult` and return that (not take it in as a parameter). The reason `MatchResult` is needed earlier is if there are no `bids` and no `asks`, it is returned. Maybe instead of returning an empty result we throw a custom `MatchException` and catch it somewhere higher up, logging the exception. 

Not sure what to do with the `tradeResult`

ngqinzhe replied:
I would suggest that `tradeResult` and its conditionals stay as it is. As for `MatchResultCreator`, we can return a `MatchResult` without taking in a `MatchResult` as a parameter by creating a new instance of it in `CreateMatchResult`.

ngqinzhe replied:
Are the test-suites supposed to have some failing tests? 

None of them throws an exception after implementing `MatchException`.

Tzadiko commented on src/OrderbookCS/MatchingAlgorithm/MatchResultUpdate.cs:
```diff
@@ -0,0 +1,20 @@
+
+namespace TradingEngineServer.Orderbook.MatchingAlgorithm
+{
+    internal class MatchResultUpdate
```
Call this class `MatchResultCreator`

Tzadiko commented on src/OrderbookCS/MatchingAlgorithm/MatchResultUpdate.cs:
```diff
@@ -0,0 +1,20 @@
+{
+    internal class MatchResultUpdate
+    {
+        public static MatchResult Update(MatchResult matchResult, TradeResult tradeResult, OrderbookEntry orderToMatchBid, OrderbookEntry orderToMatchAsk, DateTime eventTime)
```
Call this method `CreateMatchResult`. `Update` is ambiguous.

Tzadiko replied:
Call the class `MathResultCreator`

Tzadiko reviewed (changes requested):
Yes, this is better. Thanks for this submission. Please see my final comments.

Tzadiko commented:
I don’t see where you implemented MatchException. I’m assuming you haven’t pushed it.

When is MatchException thrown? It’s thrown when there are no bids and no asks. Look at the tests and see if anyone of them try to match without any bids. If there isn’t such a test, please add it. Use Assert.ThrowsException<MatchException>(…)

获取 Outlook for iOS<https://aka.ms/o0ukef>
________________________________
发件人: Qin Zhe ***@***.***>
发送时间: Tuesday, September 21, 2021 8:43:33 PM
收件人: Tzadiko/Trading-Engine ***@***.***>
抄送: Tomer Tzadok ***@***.***>; Comment ***@***.***>
主题: Re: [Tzadiko/Trading-Engine] clean up duplicate (#28)


@ngqinzhe commented on this pull request.

________________________________

In src/OrderbookCS/MatchingAlgorithm/LifoMatchingAlgorithm.cs<https://can01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FTzadiko%2FTrading-Engine%2Fpull%2F28%23discussion_r713530105&data=04%7C01%7C13tt31%40queensu.ca%7C83dc1fb9fe19404eca7a08d97d6a6375%7Cd61ecb3b38b142d582c4efb2838b925c%7C1%7C0%7C637678718165318363%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=OxehL1zmN5Mc69ICRvpx96TNnUr8eMvKv%2FRpIQEFpmk%3D&reserved=0>:

> @@ -62,11 +62,7 @@ public MatchResult Match(IEnumerable<OrderbookEntry> bids, IEnumerable<Orderbook
                 // Think of refactoring this by including it elsewhere
                 var tradeResult = TradeUtilities.CreateTradeAndFills(orderToMatchBid.Current, orderToMatchAsk.Current,


Are the test-suites supposed to have some failing tests?

None of them throws an exception after implementing MatchException.

―
You are receiving this because you commented.
Reply to this email directly, view it on GitHub<https://can01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FTzadiko%2FTrading-Engine%2Fpull%2F28%23discussion_r713530105&data=04%7C01%7C13tt31%40queensu.ca%7C83dc1fb9fe19404eca7a08d97d6a6375%7Cd61ecb3b38b142d582c4efb2838b925c%7C1%7C0%7C637678718165328317%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=9RL6rlUmbYQ0zhaIFhXV7p8P0ZZX0F6vr%2B1Gt8Gwieg%3D&reserved=0>, or unsubscribe<https://can01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAKP2F777ZJU46I5YSLYL5MLUDEYELANCNFSM5ENGD72A&data=04%7C01%7C13tt31%40queensu.ca%7C83dc1fb9fe19404eca7a08d97d6a6375%7Cd61ecb3b38b142d582c4efb2838b925c%7C1%7C0%7C637678718165328317%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=CcbtVoNvcVre2NFXYCHmYYXvJA8gWsxIhT9%2FCguqe1s%3D&reserved=0>.
Triage notifications on the go with GitHub Mobile for iOS<https://can01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fapps.apple.com%2Fapp%2Fapple-store%2Fid1477376905%3Fct%3Dnotification-email%26mt%3D8%26pt%3D524675&data=04%7C01%7C13tt31%40queensu.ca%7C83dc1fb9fe19404eca7a08d97d6a6375%7Cd61ecb3b38b142d582c4efb2838b925c%7C1%7C0%7C637678718165338276%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=LI6AuftNoOvMVBXgYt4CuuNq%2BEZ5dsxIBcaYJ%2F%2BlOs4%3D&reserved=0> or Android<https://can01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fplay.google.com%2Fstore%2Fapps%2Fdetails%3Fid%3Dcom.github.android%26referrer%3Dutm_campaign%253Dnotification-email%2526utm_medium%253Demail%2526utm_source%253Dgithub&data=04%7C01%7C13tt31%40queensu.ca%7C83dc1fb9fe19404eca7a08d97d6a6375%7Cd61ecb3b38b142d582c4efb2838b925c%7C1%7C0%7C637678718165338276%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=a4oucU7hhLHtpTN6klkRDIr6xF4hSit1sMgBlCpJ4iI%3D&reserved=0>.

ngqinzhe commented:
Yup, I just pushed the commits.

The tests for the `MatchException` is working fine.

Tzadiko commented on src/OrderbookCS/MatchingAlgorithm/FifoMatchingAlgorithm.cs:
```diff
@@ -21,7 +21,7 @@ public MatchResult Match(IEnumerable<OrderbookEntry> bids, IEnumerable<Orderbook
             var matchResult = new MatchResult();
```
Remove this. It's created lower down.

ngqinzhe replied:
We cannot return `matchResult` in the loop. I think this is necessary. 

Else we can do a simple initialisation of the variable `MatchResult matchResult;` if this is better.

Tzadiko replied:
> Else we can do a simple initialisation of the variable `MatchResult matchResult;` if this is better.

Yes, please do this. `MatchResult matchResult = null` or `MatchResult matchResult = default`.

Tzadiko commented on src/OrderbookCSTest/FifoOrderbookTests.cs:
```diff
@@ -2,6 +2,7 @@
 
 using TradingEngineServer.Instrument;
 using TradingEngineServer.Orderbook;
+using TradingEngineServer.Orderbook.MatchingAlgorithm.OrderbookIterator;
```
Is this needed?

Tzadiko reviewed (changes requested):
Left two final comments.

Tzadiko commented:
@ngqinzhe Tests failed. Please run tests before submitting code: https://github.com/Tzadiko/Trading-Engine/pull/28/checks?check_run_id=3671190735

ngqinzhe commented:
The tests I created for `MatchException` passed. 

Some other tests failed, not sure why, as I believe `MatchException` was not thrown in them, since they contained bids and asks.

Tzadiko commented:
I think it’s because some tests only add Bids or only add Asks. In that case you’ll throw on a Match(). Only throw a MatchException if there are no Bids AND no Asks. Edit the error message to reflect this new reality.

In this case, we may be returning null in the MatchingAlgorithm. I don’t want to do that. Keep the line I previously told you to delete with var matchResult = new MatchResult().

获取 Outlook for iOS<https://aka.ms/o0ukef>
________________________________
发件人: Qin Zhe ***@***.***>
发送时间: Wednesday, September 22, 2021 12:03:02 AM
收件人: Tzadiko/Trading-Engine ***@***.***>
抄送: Tomer Tzadok ***@***.***>; Comment ***@***.***>
主题: Re: [Tzadiko/Trading-Engine] clean up duplicate (#28)


The tests I created for MatchException passed.

Some other tests failed, not sure why, as I believe MatchException was not thrown in them, since they contained bids and asks.

―
You are receiving this because you commented.
Reply to this email directly, view it on GitHub<https://can01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FTzadiko%2FTrading-Engine%2Fpull%2F28%23issuecomment-924585936&data=04%7C01%7C13tt31%40queensu.ca%7C5007b5714e934964888708d97d8641b2%7Cd61ecb3b38b142d582c4efb2838b925c%7C1%7C0%7C637678837848233937%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=4vwoJfFyUOwQs2ZpSCVrYAHbE7UDwwrxWIJG5JLVz2E%3D&reserved=0>, or unsubscribe<https://can01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAKP2F72FYSXBLRQASNY44O3UDFPQNANCNFSM5ENGD72A&data=04%7C01%7C13tt31%40queensu.ca%7C5007b5714e934964888708d97d8641b2%7Cd61ecb3b38b142d582c4efb2838b925c%7C1%7C0%7C637678837848243890%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=pWw2%2BrWzc%2F%2BK3lXuABaVNNEAt474mJNo568GHV%2Fltbk%3D&reserved=0>.
Triage notifications on the go with GitHub Mobile for iOS<https://can01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fapps.apple.com%2Fapp%2Fapple-store%2Fid1477376905%3Fct%3Dnotification-email%26mt%3D8%26pt%3D524675&data=04%7C01%7C13tt31%40queensu.ca%7C5007b5714e934964888708d97d8641b2%7Cd61ecb3b38b142d582c4efb2838b925c%7C1%7C0%7C637678837848243890%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=fn5Ug2mLOY8OIsF9n3IZ1Lgxelch5dI6FZnHGTWeaiA%3D&reserved=0> or Android<https://can01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fplay.google.com%2Fstore%2Fapps%2Fdetails%3Fid%3Dcom.github.android%26referrer%3Dutm_campaign%253Dnotification-email%2526utm_medium%253Demail%2526utm_source%253Dgithub&data=04%7C01%7C13tt31%40queensu.ca%7C5007b5714e934964888708d97d8641b2%7Cd61ecb3b38b142d582c4efb2838b925c%7C1%7C0%7C637678837848243890%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=PN9GsW8oty3bxOatehtP42rC%2BvAB65K9ucZGP7kERy0%3D&reserved=0>.

Tzadiko commented:
In other words check for no bids AND no asks. Throw if that condition is met. Then check for No Bids OR No Asks, return empty match result in that case. Reinstate  car MatchResult = new MatchResult() at top of method

获取 Outlook for iOS<https://aka.ms/o0ukef>
________________________________
发件人: Tomer Tzadok ***@***.***>
发送时间: Wednesday, September 22, 2021 12:07:42 AM
收件人: Tzadiko/Trading-Engine ***@***.***>; Tzadiko/Trading-Engine ***@***.***>
抄送: Comment ***@***.***>
主题: Re: [Tzadiko/Trading-Engine] clean up duplicate (#28)

I think it’s because some tests only add Bids or only add Asks. In that case you’ll throw on a Match(). Only throw a MatchException if there are no Bids AND no Asks. Edit the error message to reflect this new reality.

In this case, we may be returning null in the MatchingAlgorithm. I don’t want to do that. Keep the line I previously told you to delete with var matchResult = new MatchResult().

获取 Outlook for iOS<https://aka.ms/o0ukef>
________________________________
发件人: Qin Zhe ***@***.***>
发送时间: Wednesday, September 22, 2021 12:03:02 AM
收件人: Tzadiko/Trading-Engine ***@***.***>
抄送: Tomer Tzadok ***@***.***>; Comment ***@***.***>
主题: Re: [Tzadiko/Trading-Engine] clean up duplicate (#28)


The tests I created for MatchException passed.

Some other tests failed, not sure why, as I believe MatchException was not thrown in them, since they contained bids and asks.

―
You are receiving this because you commented.
Reply to this email directly, view it on GitHub<https://can01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FTzadiko%2FTrading-Engine%2Fpull%2F28%23issuecomment-924585936&data=04%7C01%7C13tt31%40queensu.ca%7C5007b5714e934964888708d97d8641b2%7Cd61ecb3b38b142d582c4efb2838b925c%7C1%7C0%7C637678837848233937%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=4vwoJfFyUOwQs2ZpSCVrYAHbE7UDwwrxWIJG5JLVz2E%3D&reserved=0>, or unsubscribe<https://can01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAKP2F72FYSXBLRQASNY44O3UDFPQNANCNFSM5ENGD72A&data=04%7C01%7C13tt31%40queensu.ca%7C5007b5714e934964888708d97d8641b2%7Cd61ecb3b38b142d582c4efb2838b925c%7C1%7C0%7C637678837848243890%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=pWw2%2BrWzc%2F%2BK3lXuABaVNNEAt474mJNo568GHV%2Fltbk%3D&reserved=0>.
Triage notifications on the go with GitHub Mobile for iOS<https://can01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fapps.apple.com%2Fapp%2Fapple-store%2Fid1477376905%3Fct%3Dnotification-email%26mt%3D8%26pt%3D524675&data=04%7C01%7C13tt31%40queensu.ca%7C5007b5714e934964888708d97d8641b2%7Cd61ecb3b38b142d582c4efb2838b925c%7C1%7C0%7C637678837848243890%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=fn5Ug2mLOY8OIsF9n3IZ1Lgxelch5dI6FZnHGTWeaiA%3D&reserved=0> or Android<https://can01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fplay.google.com%2Fstore%2Fapps%2Fdetails%3Fid%3Dcom.github.android%26referrer%3Dutm_campaign%253Dnotification-email%2526utm_medium%253Demail%2526utm_source%253Dgithub&data=04%7C01%7C13tt31%40queensu.ca%7C5007b5714e934964888708d97d8641b2%7Cd61ecb3b38b142d582c4efb2838b925c%7C1%7C0%7C637678837848243890%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=PN9GsW8oty3bxOatehtP42rC%2BvAB65K9ucZGP7kERy0%3D&reserved=0>.

ngqinzhe commented:
Tests are still failing after implementing your solution. 

The failed tests are mainly regarding the `matchResult.Fills.Count`.

Tzadiko commented:
> Tests are still failing after implementing your solution.
> 
> The failed tests are mainly regarding the `matchResult.Fills.Count`.

Try to figure out what broke, otherwise I'll take a look later tonight. Please make sure this MR reflects your branch before I take a look at it. It's hard to understand what is going on when your local branch is more up-to-date that your remote branch, as we don't have a uniform view of the code.

ngqinzhe commented:
Just updated the repo. 

do `matchResult.Fills.Count` increment each time a bid or ask is added to the orderbook, or each time a match happens?

Tzadiko commented:
Match

获取 Outlook for iOS<https://aka.ms/o0ukef>

Tzadiko commented on src/OrderEntryProtoCS/OrderEntryProtoCS.csproj:
```diff
@@ -7,10 +7,6 @@
   </PropertyGroup>
 
-  <ItemGroup>
-    <None Remove="OrderEntryCommunication.proto" />
```
Why is this change being made in this MR?

Tzadiko commented on src/OrderEntryProtoCS/OrderEntryProtoCS.csproj:
```diff
@@ -20,8 +16,4 @@
         </PackageReference>
     </ItemGroup>
 
-    <ItemGroup>
```
Why is this change being made in this MR?

Tzadiko reviewed (changes requested):
Unsure why this is added

ngqinzhe commented:
If that's the case , in `LifoOrderbook_MatchThreeOrders_AskRemainsResting` for example,
```
var askOrder = new Order(new OrderCore(askOrderOrderId, string.Empty, 0), askPrice, 100, false); // Masive ask wall.
var buyOrder = new Order(new OrderCore(buyOrderOrderId, string.Empty, 0), 10_001, 15, true);
var buyOrder2 = new Order(new OrderCore(secondbuyOrderOrderId, string.Empty, 0), 10_001, 15, true);

LifoOrderbook.AddOrder(askOrder);
LifoOrderbook.AddOrder(buyOrder);
LifoOrderbook.AddOrder(buyOrder2);

var matchResult = LifoOrderbook.Match();
var spread = LifoOrderbook.GetSpread();

Assert.AreEqual(4, matchResult.Fills.Count);
```

The number of matches should only be 2 instead of 4 no? Since the two buy orders match with the sell order.

Tzadiko commented on src/OrderbookCS/MatchingAlgorithm/FifoMatchingAlgorithm.cs:
```diff
@@ -56,11 +58,7 @@ public MatchResult Match(IEnumerable<OrderbookEntry> bids, IEnumerable<Orderbook
-                Limit relevantOrderbookLimit = buySideIsAggressor ? orderToMatchAsk.ParentLimit : orderToMatchBid.ParentLimit;
-                var orderbookUpdate = OrderbookUtilities.CreateIncrementalOrderbookUpdate(relevantOrderbookLimit, eventTime);
-                matchResult.AddIncrementalOrderbookUpdate(orderbookUpdate);
+                matchResult = MatchResultCreator.CreateMatchResult(tradeResult, orderToMatchBid, orderToMatchAsk, eventTime);
```
This is the problem with your code. `matchResult` is being replaced every single time. Add a `Merge(MatchResult other)` method on `MatchResult`

```csharp
matchResult.Merge(MatchResultCreator.CreateMatchResult(tradeResult, orderToMatchBid, orderToMatchAsk, eventTime));
```

ngqinzhe replied:
Yes. Would changing to `MatchResultCreator.Update` but passing in `matchResult` as a parameter be a better approach? Feels more readable compared to implementing `Merge`.

Tzadiko replied:
Sure. Change the name to `UpdateMatchResult`

Tzadiko commented:
> If that's the case , in `LifoOrderbook_MatchThreeOrders_AskRemainsResting` for example,
> 
> ```
> var askOrder = new Order(new OrderCore(askOrderOrderId, string.Empty, 0), askPrice, 100, false); // Masive ask wall.
> var buyOrder = new Order(new OrderCore(buyOrderOrderId, string.Empty, 0), 10_001, 15, true);
> var buyOrder2 = new Order(new OrderCore(secondbuyOrderOrderId, string.Empty, 0), 10_001, 15, true);
> 
> LifoOrderbook.AddOrder(askOrder);
> LifoOrderbook.AddOrder(buyOrder);
> LifoOrderbook.AddOrder(buyOrder2);
> 
> var matchResult = LifoOrderbook.Match();
> var spread = LifoOrderbook.GetSpread();
> 
> Assert.AreEqual(4, matchResult.Fills.Count);
> ```
> 
> The number of matches should only be 2 instead of 4 no? Since the two buy orders match with the sell order.

A fill is generated for each side of the trade. There are two orders matching against one ask wall. The first order generates two, the second generates two, for a total of 4.

Tzadiko commented on src/OrderbookCS/MatchingAlgorithm/FifoMatchingAlgorithm.cs:
```diff
@@ -58,8 +58,7 @@ public MatchResult Match(IEnumerable<OrderbookEntry> bids, IEnumerable<Orderbook
                     fillQuantity, AllocationAlgorithm.Fifo, eventTime);
-                matchResult = MatchResultCreator.CreateMatchResult(tradeResult, orderToMatchBid, orderToMatchAsk, eventTime);
-
+                matchResult = MatchResultCreator.UpdateMatchResult(matchResult, tradeResult, orderToMatchBid, orderToMatchAsk, eventTime);
```
If the method is called update, don't return `MatchResult`

ngqinzhe replied:
done

ngqinzhe commented:
Added the relevant changes. Tests passed. 

For `src/OrderEntryProtoCS/OrderEntryProtoCS.csproj` I think I accidentally touched it. I have replaced it with the one in your repo.

Tzadiko commented:
> Added the relevant changes. Tests passed.
> 
> For `src/OrderEntryProtoCS/OrderEntryProtoCS.csproj` I think I accidentally touched it. I have replaced it with the one in your repo.

Please address this: https://github.com/Tzadiko/Trading-Engine/pull/28#discussion_r714039024

Then I'll merge.

Tzadiko commented:
@ngqinzhe Thanks for your work here. Feel free to join the Discord: https://discord.gg/m3rhCua6

ngqinzhe commented:
Np, thanks for the feedbacks. Learned a lot
